common/src/app/common/geom/shapes/flex_layout/drop_area.cljc

```diff
@@ -0,0 +1,169 @@
+                  (let [half-point-height (+ (- box-y y) (/ box-height 2))]
+                    [(-> (gsr/make-rect x y width half-point-height)
+                         (assoc :index (if reverse? (inc index) index)))
+                     (-> (gsr/make-rect x (+ y half-point-height) width (- height half-point-height))
+                         (assoc :index (if reverse? index (inc index))))]))
+
+                result (conj result line-area-1 line-area-2)]
+
+            [result parent-rect (+ x width) (+ y height)]))
+
+        redfn-lines
+        (fn [[result from-idx prev-x prev-y] [{:keys [start-p layout-gap-row layout-gap-col num-children line-width line-height]} next]]
+          (let [start-p (gmt/transform-point-center start-p (gco/center-shape frame) (:transform-inverse frame))
+
+                prev-x (or prev-x (:x frame))
+                prev-y (or prev-y (:y frame))
+                last? (nil? next)
+
+                line-width
+                (if row?
+                  (:width frame)
+                  (+ line-width margin-x
+                     (if row? (* layout-gap-row (dec num-children)) 0)))
+
+                line-height
+                (if col?
+                  (:height frame)
+                  (+ line-height margin-y
+                     (if col?
+                       (* layout-gap-col (dec num-children))
+                       0)))
+
+                box-x
+                (- (:x start-p)
+                   (cond
+                     h-center? (/ line-width 2)
+                     h-end? line-width
+                     :else 0))
+
+                box-y
+                (- (:y start-p)
+                   (cond
+                     v-center? (/ line-height 2)
+                     v-end? line-height
+                     :else 0))
+
+                x (if row? (:x frame) prev-x)
+                y (if col? (:y frame) prev-y)
+
+                width (cond
+                        (and col? last?)
+                        (- (+ (:x frame) (:width frame)) x)
+
+                        row?
+                        (:width frame)
+
+                        :else
+                        (+ line-width (- box-x prev-x) (/ layout-gap-row 2)))
+
+                height (cond
+                         (and row? last?)
+                         (- (+ (:y frame) (:height frame)) y)
+
+                         col?
+                         (:height frame)
+
+                         :else
+                         (+ line-height (- box-y prev-y) (/ layout-gap-col 2)))
+
+                line-area (gsr/make-rect x y width height)
+
+                children (subvec children from-idx (+ from-idx num-children))
+
+                result (first (reduce redfn-child [result line-area] (d/with-next children)))]
+
+            [result (+ from-idx num-children) (+ x width) (+ y height)]))]
+
+    (first (reduce redfn-lines [[] 0] (d/with-next (:layout-lines layout-data))))))
```

Comment:
I suggest consider using `loop` here instead of `reduce`. I guess this will improve readability and performance